server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2616,6 +2618,60 @@ public void removePeerRecoveryRetentionLease(String nodeId, ActionListener<Repli
         replicationTracker.removePeerRecoveryRetentionLease(nodeId, listener);
     }
 
+    /**
+     * Determines a balance between file-based and operations-based peer recoveries. The number of operations that will be used in an
+     * operations-based peer recovery is limited to this proportion of the total number of documents in the shard (including deleted
+     * documents) on the grounds that a file-based peer recovery may copy all of the documents in the shard over to the new peer, but is
+     * significantly faster than replaying the missing operations on the peer, so once a peer falls far enough behind the primary it makes
+     * more sense to copy all the data over again instead of replaying history.
+     *
+     * Defaults to retaining history for up to 10% of the documents in the shard. This can only be changed in tests, since this setting is
+     * intentionally unregistered.
+     */
+    public static final Setting<Double> REASONABLE_OPERATIONS_BASED_RECOVERY_PROPORTION_SETTING
+        = Setting.doubleSetting("index.recovery.reasonable_operations_based_recovery_proportion", 0.1, 0.0,
+            Setting.Property.Dynamic, Setting.Property.IndexScope);
+
+    private LongSupplier getMinimumReasonableRetainedSeqNoSupplier(final IndexSettings indexSettings) {
+        return new LongSupplier() {
+
+            private long generation = Long.MIN_VALUE;
+            private long firstReasonableSeqNo = Long.MIN_VALUE;
+            private final double reasonableOperationsBasedRecoveryProportion
+                = REASONABLE_OPERATIONS_BASED_RECOVERY_PROPORTION_SETTING.get(indexSettings.getSettings());
+
+            @Override
+            public synchronized long getAsLong() {
+                try (Engine.IndexCommitRef safeCommitRef = getEngine().acquireSafeIndexCommit()) {
```

Comment:
Please ignore this as it is not correct. We won't revisit the deletion policy on releasing a commit unless the safe commit advances.

Reply:
I too am a little uneasy with the costs of this. With this implementation each call to `getAsLong()` is `synchronized` and obtains a new commit ref, which might occasionally be expensive to release. It only does this on replication groups that aren't fully-assigned, of course. An alternative would be to recalculate this while advancing the safe commit, then this could just become a volatile read. I couldn't determine a neat way to hook this calculation in the right place. Any ideas?

Reply:
I tried moving the calculation to `CombinedDeletionPolicy#updateRetentionPolicy` in c220530174d6e4a41b7d473ce1efd303dbe09fe1. I think it makes sense there. WDYT?